save-orchestrator/src/main/kotlin/com/saveourtool/save/orchestrator/docker/DockerAgentRunner.kt

```diff
@@ -0,0 +1,196 @@
+            dockerClient.startContainerCmd(agentId).exec()
+        }
+    }
+
+    override fun stop(executionId: String) {
+        val runningContainersForExecution = dockerClient.listContainersCmd().withStatusFilter(listOf("running")).exec()
+            .filter { container -> container.names.any { it.contains("-$executionId-") } }
+        runningContainersForExecution.map { it.id }.forEach { agentId ->
+            dockerClient.stopContainerCmd(agentId).exec()
+        }
+    }
+
+    override fun stopByAgentId(agentId: String) {
+        logger.info("Stopping agent with id=$agentId")
+        val state = dockerClient.inspectContainerCmd(agentId).exec().state
+        if (state.running == true) {
+            dockerClient.stopContainerCmd(agentId).exec()
+            logger.info("Agent with id=$agentId has been stopped")
+        } else {
+            logger.warn("Agent with id=$agentId was requested to be stopped, but it actual state=$state")
+        }
+    }
+
+    override fun cleanup(executionId: Long) {
+        val containersForExecution = dockerClient.listContainersCmd().withNameFilter(listOf("-$executionId-")).exec()
+
+        containersForExecution.map { it.id }.forEach { containerId ->
+            logger.info("Removing container $containerId")
+            val existingContainerIds = dockerClient.listContainersCmd().withShowAll(true).exec()
+                .map {
+                    it.id
+                }
+            if (containerId in existingContainerIds) {
+                dockerClient.removeContainerCmd(containerId).exec()
+            } else {
+                logger.info("Container $containerId is not present, so won't attempt to remove")
+            }
+        }
+    }
+
+    /**
+     * Creates a docker container
+     *
+     * @param runCmd an entrypoint for docker container with CLI arguments
+     * @param containerName a name for the created container
+     * @param baseImageId id of the base docker image for this container
+     * @param workingDir working directory for [runCmd]
+     * @return id of created container or null if it wasn't created
+     * @throws DockerException if docker daemon has returned an error
+     * @throws RuntimeException if an exception not specific to docker has occurred
+     */
+    @Suppress("UnsafeCallOnNullableType", "TOO_LONG_FUNCTION")
+    private fun createContainerFromImage(baseImageId: String,
+                                         workingDir: String,
+                                         runCmd: String,
+                                         containerName: String,
+    ): String {
+        val baseImage = dockerClient.listImagesCmd().execTimed(meterRegistry, "$DOCKER_METRIC_PREFIX.image.list")!!.find {
+            // fixme: sometimes createImageCmd returns short id without prefix, sometimes full and with prefix.
+            it.id.replaceFirst("sha256:", "").startsWith(baseImageId.replaceFirst("sha256:", ""))
```

Comment:
why replace, if it is the same in both IDs: `sha256:`

Reply:
I don't really remember, it's another piece of old code. I think, we can get one id with this prefix and the other one without and I couldn't fully determine why. I think, it might behave differently on different versions of Docker. So if any of IDs has this prefix, we compare only stuff after it